### PR TITLE
remove redundant remote_user for local setting

### LINKED ISSUE #59642: Unsafe use of fork() after join()-ing threads causes ansible to hang under certain conditions.
##### SUMMARY

Under certain conditions (see "STEPS TO REPRODUCE"), an ansible playbook run with local connection will hang when gathering facts.

We have traced the problem to unsafe use of `fork()` (via creation of `WorkerProcess`[1]) after result `Threads` have been `join()`ed. [2]. The python docs [3] warn against this:

> Note that safely forking a multithreaded process is problematic

This bug report [4] makes the point somewhat more emphatically:

> Do realize that while working on this it is fundamentally *impossible* per POSIX for os.fork() to be safely used at the Python level in a process also using pthreads.  That this _ever_ appeared to work is a pure accident of implementations of underlying libc, malloc, system libraries, and kernel behaviors.  POSIX considers it undefined behavior.  Nothing done in CPython can avoid that.  Any "fix" for these kinds of issues is merely working around the inevitable which will re-occur.

We have experimented with using `forkserver` instead of `fork` for starting subprocesses in ansible, and, barring some minor workaround for serialization issues that arise as a result, this seems to fix the problem. We would therefore propose either a configuration option which allowed users to pick a start method or - perhaps more correctly/definitively but - make this the default where available (though we realise that there could be some consequences here). We would appreciate some opinions/guidance on whether this is a sane approach before we dive in and try to make the change.

[1] https://github.com/ansible/ansible/blob/v2.7.0/lib/ansible/plugins/strategy/__init__.py#L320
[2] https://github.com/ansible/ansible/blob/v2.7.0/lib/ansible/plugins/strategy/__init__.py#L228
[3] https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods
[4] https://bugs.python.org/issue35866#msg340442

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
Core

##### ANSIBLE VERSION
```
ansible 2.7.0
  config file = None
  configured module search path = ['/home/tw-admin/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.8 (default, Jan 14 2019, 11:02:34) [GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
```

##### OS / ENVIRONMENT
```
$ lsb_release -d
Description:	Ubuntu 18.04.2 LTS

$ uname -a
Linux amp 4.15.0-1044-aws #46-Ubuntu SMP Thu Jul 4 13:38:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

$ dpkg -l | grep libc6
ii  libc6:amd64                    2.27-3ubuntu1                     amd64        GNU C Library: Shared libraries
ii  libc6-dbg:amd64                2.27-3ubuntu1                     amd64        GNU C Library: detached debugging symbols

$ dpkg -l | grep libgcc
ii  libgcc1:amd64                  1:8.3.0-6ubuntu1~18.04.1          amd64        GCC support library
```

##### STEPS TO REPRODUCE

It's complicated. The issue arises when we invoke `systemd` or `command` to both enable and start on filebeat service.

e.g. 

```
---
  - command: systemctl enable filebeat.service
  - command: systemctl start filebeat.service
```

and then try and run any other playbook.

##### EXPECTED RESULTS

The second playbook completes.

##### ACTUAL RESULTS

The second playbook hangs while gathering facts.

TL; DR: all of the results below are expected (or maybe "allowed") behaviour from CPython's point of view. Usually we get away with it in ansible because stack unwinding is very quick, but something about our setup makes this slower than usual.

Backtrace from `gdb` shows that the worker process is hung around here:

```
  File "/usr/local/lib/python3.6/dist-packages/ansible/playbook/play_context.py", line 433, in set_task_and_variable_override
    new_info.remote_user = pwd.getpwuid(os.getuid()).pw_name
  File "/usr/local/lib/python3.6/dist-packages/ansible/executor/task_executor.py", line 465, in _execute
    self._play_context = self._play_context.set_task_and_variable_override(task=self._task, variables=variables, templar=templar)
```

The call to `getpwuid` requires opening of a shared library, `libnss_compat.so.2`, which requires a lock, which claims to be held by another process...

```
#0  __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
#1  0x00007f56aa26c098 in __GI___pthread_mutex_lock (mutex=0x7f56aaa9a990 <_rtld_global+2352>) at ../nptl/pthread_mutex_lock.c:113
#2  0x00007f56aa87e01e in _dl_add_to_namespace_list (new=0x1fd7f30, nsid=0) at dl-object.c:33
#3  0x00007f56aa8793c7 in _dl_map_object_from_fd (name=name@entry=0x7ffc3e1e3f90 "libnss_compat.so.2", origname=origname@entry=0x0, fd=-1, fbp=fbp@entry=0x7ffc3e1e37b0, realname=<optimised out>, 
    loader=loader@entry=0x0, l_type=<optimised out>, mode=<optimised out>, stack_endp=<optimised out>, nsid=<optimised out>) at dl-load.c:1322
#4  0x00007f56aa87affc in _dl_map_object (loader=0x0, loader@entry=0x7f56aaa90000, name=name@entry=0x7ffc3e1e3f90 "libnss_compat.so.2", type=type@entry=2, trace_mode=trace_mode@entry=0, 
    mode=mode@entry=-1879048191, nsid=<optimised out>) at dl-load.c:2389
#5  0x00007f56aa886ee4 in dl_open_worker (a=a@entry=0x7ffc3e1e3d20) at dl-open.c:235
#6  0x00007f56aa5e82df in __GI__dl_catch_exception (exception=exception@entry=0x7ffc3e1e3d00, operate=operate@entry=0x7f56aa886dc0 <dl_open_worker>, args=args@entry=0x7ffc3e1e3d20) at dl-error-skeleton.c:196
#7  0x00007f56aa8867ca in _dl_open (file=0x7ffc3e1e3f90 "libnss_compat.so.2", mode=-2147483647, caller_dlopen=0x7f56aa5c9886 <nss_load_library+294>, nsid=<optimised out>, argc=8, argv=<optimised out>, 
    env=0x7ffc3e1e6e00) at dl-open.c:605
#8  0x00007f56aa5e73ad in do_dlopen (ptr=ptr@entry=0x7ffc3e1e3f50) at dl-libc.c:96
```

however, that process is no longer running by this point. Tracing it back by catching `clone()` system calls, we determined that the holding process was actually an ansible results thread. At the point of the fork, the backtrace of the results thread (or rather what's left of the underlying LWP - by this point `join()` has returned and it's Python C code that is cleaning up the pthread) looks like this:

```
#0  _Unwind_IteratePhdrCallback (info=<optimised out>, size=<optimised out>, ptr=0x7fffefb0e970) at ../../../src/libgcc/unwind-dw2-fde-dip.c:406
#1  0x00007ffff7b49f21 in __GI___dl_iterate_phdr (callback=callback@entry=0x7fffee907890 <_Unwind_IteratePhdrCallback>, data=data@entry=0x7fffefb0e970) at dl-iteratephdr.c:75
#2  0x00007fffee908b31 in _Unwind_Find_FDE (pc=0x6383c4 <PyThread_exit_thread+20>, bases=bases@entry=0x7fffefb0edc8) at ../../../src/libgcc/unwind-dw2-fde-dip.c:469
#3  0x00007fffee905223 in uw_frame_state_for (context=context@entry=0x7fffefb0ed20, fs=fs@entry=0x7fffefb0ea60) at ../../../src/libgcc/unwind-dw2.c:1257
#4  0x00007fffee90672f in _Unwind_ForcedUnwind_Phase2 (exc=exc@entry=0x7fffefb0fd70, context=context@entry=0x7fffefb0ed20, frames_p=frames_p@entry=0x7fffefb0ec28) at ../../../src/libgcc/unwind.inc:193
#5  0x00007fffee906cfc in _Unwind_ForcedUnwind (exc=0x7fffefb0fd70, stop=stop@entry=0x7ffff77d5d80 <unwind_stop>, stop_argument=<optimised out>) at ../../../src/libgcc/unwind.inc:217
#6  0x00007ffff77d5f10 in __GI___pthread_unwind (buf=<optimised out>) at unwind.c:121
#7  0x00007ffff77cdae5 in __do_cancel () at pthreadP.h:297
#8  __pthread_exit (value=<optimised out>) at pthread_exit.c:28
#9  0x00007ffff7b14504 in __pthread_exit (retval=<optimised out>) at forward.c:173
#10 0x00000000006383c5 in PyThread_exit_thread () at ../Python/thread_pthread.h:300
#11 0x00000000005e5f0f in t_bootstrap () at ../Modules/_threadmodule.c:1030
#12 0x0000000000638084 in pythread_wrapper (arg=<optimised out>) at ../Python/thread_pthread.h:205
#13 0x00007ffff77cc6db in start_thread (arg=0x7fffefb0f700) at pthread_create.c:463
#14 0x00007ffff7b0588f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

The call from `_dl_iterate_phdr` requires the same lock that `_dlopen` is waiting for in the first backtrace. From the worker process' point of view, that lock is never released.

Something about our particular setup means that the stack unwinding takes an unusually long time. We only see the problem when there is a cold page cache, which seems to make sense when you look at what `_Unwind_IteratePhdrCallback` does (tl;dr: it requires a lot of IO). We also only see it on certain EC2 instance types - it seems to be about balance between IO and compute.

### PR DESCRIPTION
**Pull request by bcoca**
local action plugin already does and this also should fix
  fork/thread issue by removing use of pwd library

  fixes #59642

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

play_context

### CODE REVIEW

**Comment by felixfontein** [comment]
This PR breaks the following playbook (`t.yml`) for me:
```
- hosts: remote_server
  tasks:
  - copy:
      src: t.yml
      dest: /tmp/t.yml
    delegate_to: localhost
```
It does not break when `hosts: localhost`.

**Comment by felixfontein** [comment]
```
fatal: [remote-server]: UNREACHABLE! => {"changed": false, "msg": "Failed to create temporary directory.In some cases, you may have been able to authenticate and did not have permissions on the target directory. Consider changing the remote tmp path in ansible.cfg to a path rooted in \"/tmp\", for more error information use -vvv. Failed command was: ( umask 77 && mkdir -p \"` echo /root/.ansible/tmp `\"&& mkdir \"` echo /root/.ansible/tmp/ansible-tmp-1604856819.0247443-194052-180695334418942 `\" && echo ansible-tmp-1604856819.0247443-194052-180695334418942=\"` echo /root/.ansible/tmp/ansible-tmp-1604856819.0247443-194052-180695334418942 `\" ), exited with result 1", "unreachable": true}
```

**Comment by bcoca** [comment]
I cannot reproduce with that play.

**Comment by felixfontein** [comment]
Looks like this only happens (for me) if the remote user is `root`, while I run the playbook as a regular user.
